File: DynamicKey/AgoraDynamicKey/nodejs/src/ApaasTokenBuilder.js
const md5 = require('md5')

const AccessToken = require('./AccessToken2').AccessToken2
const ServiceApaas = require('./AccessToken2').ServiceApaas
const ServiceChat = require('./AccessToken2').ServiceChat
const ServiceRtm = require('./AccessToken2').ServiceRtm

class ApaasTokenBuilder {
    /**
     * build user room token
     * @param appId             The App ID issued to you by Agora. Apply for a new App ID from
     *                          Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param appCertificate    Certificate of the application that you registered in
     *                          the Agora Dashboard. See Get an App Certificate.
     * @param roomUuid          The room's id, must be unique.
     * @param userUuid          The user's id, must be unique.
     * @param role              The user's role.
     * @param expire            represented by the number of seconds elapsed since now. If, for example, you want to access the
     *                          Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
     * @return The user room token.
     */
    static buildRoomUserToken(appId, appCertificate, roomUuid, userUuid, role, expire) {
        let accessToken = new AccessToken(appId, appCertificate, 0, expire)

        let chatUserId = md5(userUuid)
        let apaasService = new ServiceApaas(roomUuid, userUuid, role)
        accessToken.add_service(apaasService)

        let rtmService = new ServiceRtm(userUuid)
        rtmService.add_privilege(ServiceRtm.kPrivilegeLogin, expire)
        accessToken.add_service(rtmService)

        let chatService = new ServiceChat(chatUserId)
        chatService.add_privilege(ServiceChat.kPrivilegeUser, expire)
        accessToken.add_service(chatService)

        return accessToken.build()
    }

    /**
     * build user token
     * @param appId             The App ID issued to you by Agora. Apply for a new App ID from
     *                          Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param appCertificate    Certificate of the application that you registered in
     *                          the Agora Dashboard. See Get an App Certificate.
     * @param userUuid          The user's id, must be unique.
     * @param expire            represented by the number of seconds elapsed since now. If, for example, you want to access the
     *                          Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
     * @return The user token.
     */
    static buildUserToken(appId, appCertificate, userUuid, expire) {
        let accessToken = new AccessToken(appId, appCertificate, 0, expire)
        let apaasService = new ServiceApaas('', userUuid)
        apaasService.add_privilege(ServiceApaas.PRIVILEGE_USER, expire)
        accessToken.add_service(apaasService)

        return accessToken.build()
    }

    /**
     * build app token
     * @param appId          The App ID issued to you by Agora. Apply for a new App ID from
     *                       Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param appCertificate Certificate of the application that you registered in
     *                       the Agora Dashboard. See Get an App Certificate.
     * @param expire         represented by the number of seconds elapsed since now. If, for example, you want to access the
     *                       Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
     * @return The app token.
     */
    static buildAppToken(appId, appCertificate, expire) {
        let accessToken = new AccessToken(appId, appCertificate, 0, expire)
        let apaasService = new ServiceApaas()
        apaasService.add_privilege(ServiceApaas.PRIVILEGE_APP, expire)
        accessToken.add_service(apaasService)

        return accessToken.build()
    }
}

module.exports.ApaasTokenBuilder = ApaasTokenBuilder

File: DynamicKey/AgoraDynamicKey/dart/lib/src/apaas_token_builder.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

import 'access_token.dart';

/// Builds Token007 tokens for APaaS rooms, users, and applications.
class ApaasTokenBuilder {
  /// Builds an APaaS room-user token.
  ///
  /// [appId] is the App ID issued by Agora; create one in Agora Dashboard if needed.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [roomUuid] is the room ID and must be unique.
  /// [userUuid] is the user ID and must be unique.
  /// [role] is the user's role.
  /// [expireSeconds] is the expiration time in seconds from now; use `600` for 10 minutes.
  ///
  /// Returns the generated room-user token.
  static String buildRoomUserToken({
    required String appId,
    required String appCertificate,
    required String roomUuid,
    required String userUuid,
    required int role,
    required int expireSeconds,
  }) {
    final token = AccessToken.create(
      appId,
      appCertificate,
      expire: expireSeconds,
    );
    token
      ..addService(
        ServiceApaas(roomUuid, userUuid, role)
          ..addPrivilege(ServiceApaas.privilegeRoomUser, expireSeconds),
      )
      ..addService(
        ServiceRtm(userUuid)
          ..addPrivilege(ServiceRtm.privilegeLogin, expireSeconds),
      )
      ..addService(
        ServiceChat(md5.convert(utf8.encode(userUuid)).toString())
          ..addPrivilege(ServiceChat.privilegeUser, expireSeconds),
      );
    return token.build();
  }

  /// Builds an APaaS user token.
  ///
  /// [appId] is the App ID issued by Agora; create one in Agora Dashboard if needed.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [userUuid] is the user ID and must be unique.
  /// [expireSeconds] is the expiration time in seconds from now; use `600` for 10 minutes.
  ///
  /// Returns the generated user token.
  static String buildUserToken({
    required String appId,
    required String appCertificate,
    required String userUuid,
    required int expireSeconds,
  }) {
    final token = AccessToken.create(
      appId,
      appCertificate,
      expire: expireSeconds,
    );
    token.addService(
      ServiceApaas('', userUuid, -1)
        ..addPrivilege(ServiceApaas.privilegeUser, expireSeconds),
    );
    return token.build();
  }

  /// Builds an APaaS application token.
  ///
  /// [appId] is the App ID issued by Agora; create one in Agora Dashboard if needed.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [expireSeconds] is the expiration time in seconds from now; use `600` for 10 minutes.
  ///
  /// Returns the generated application token.
  static String buildAppToken({
    required String appId,
    required String appCertificate,
    required int expireSeconds,
  }) {
    final token = AccessToken.create(
      appId,
      appCertificate,
      expire: expireSeconds,
    );
    token.addService(
      ServiceApaas()..addPrivilege(ServiceApaas.privilegeApp, expireSeconds),
    );
    return token.build();
  }
}
